From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ashish Mhetre <amhetre@nvidia.com>
Date: Fri, 21 Nov 2025 08:31:36 +0000
Subject: [PATCH 082/231] NVIDIA: SAUCE: iommu/arm-smmu-v3: Restore IRQs on
 resume

The ARM SMMU v3 driver was not restoring interrupt configuration after
resume from suspend, leaving the device unable to generate interrupts
for events, errors, and page requests. This resulted in silent failures
where DMA faults and other critical events went unreported after a
suspend/resume cycle.
Add a new arm_smmu_enable_irqs() function that re-enables interrupt
generation in the SMMU hardware by writing to the IRQ_CTRL register.
This function is separate from arm_smmu_setup_irqs() which registers
IRQ handlers using devm_request_irq() and should only be called during
probe. The IRQ handlers themselves persist through suspend/resume as
device-managed resources, so only the hardware enable bits need to be
restored.
Call arm_smmu_enable_irqs() in arm_smmu_runtime_resume() after the
device reset to ensure interrupts are properly re-enabled on resume.

Bug 5088995

Change-Id: Iea5e5afa0c92ba762255c8a533fc5c05d9858e08
Signed-off-by: Ashish Mhetre <amhetre@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3496353
Reviewed-by: Pritesh Raithatha <praithatha@nvidia.com>
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
Tested-by: Ishan Shah <ishah@nvidia.com>
Reviewed-by: svcacv <svcacv@nvidia.com>
(cherry picked from commit bd85e3c4db26954c06cd73c9452a73a17a6d8ae5)

diff --git a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
index 43174846a84b3..05f4816e7bca7 100644
--- a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
+++ b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
@@ -4663,10 +4663,26 @@ static void arm_smmu_setup_unique_irqs(struct arm_smmu_device *smmu)
 	}
 }
 
+static int arm_smmu_enable_irqs(struct arm_smmu_device *smmu)
+{
+	int ret = 0;
+	u32 irqen_flags = IRQ_CTRL_EVTQ_IRQEN | IRQ_CTRL_GERROR_IRQEN;
+
+	if (smmu->features & ARM_SMMU_FEAT_PRI)
+		irqen_flags |= IRQ_CTRL_PRIQ_IRQEN;
+
+	/* Enable interrupt generation on the SMMU */
+	ret = arm_smmu_write_reg_sync(smmu, irqen_flags,
+				      ARM_SMMU_IRQ_CTRL, ARM_SMMU_IRQ_CTRLACK);
+	if (ret)
+		dev_warn(smmu->dev, "failed to enable irqs\n");
+
+	return ret;
+}
+
 static int arm_smmu_setup_irqs(struct arm_smmu_device *smmu)
 {
 	int ret, irq;
-	u32 irqen_flags = IRQ_CTRL_EVTQ_IRQEN | IRQ_CTRL_GERROR_IRQEN;
 
 	/* Disable IRQs first */
 	ret = arm_smmu_write_reg_sync(smmu, 0, ARM_SMMU_IRQ_CTRL,
@@ -4692,16 +4708,7 @@ static int arm_smmu_setup_irqs(struct arm_smmu_device *smmu)
 	} else
 		arm_smmu_setup_unique_irqs(smmu);
 
-	if (smmu->features & ARM_SMMU_FEAT_PRI)
-		irqen_flags |= IRQ_CTRL_PRIQ_IRQEN;
-
-	/* Enable interrupt generation on the SMMU */
-	ret = arm_smmu_write_reg_sync(smmu, irqen_flags,
-				      ARM_SMMU_IRQ_CTRL, ARM_SMMU_IRQ_CTRLACK);
-	if (ret)
-		dev_warn(smmu->dev, "failed to enable irqs\n");
-
-	return 0;
+	return arm_smmu_enable_irqs(smmu);
 }
 
 static int arm_smmu_device_disable(struct arm_smmu_device *smmu)
@@ -5540,9 +5547,18 @@ static void arm_smmu_driver_unregister(struct platform_driver *drv)
 static int __maybe_unused arm_smmu_runtime_resume(struct device *dev)
 {
 	struct arm_smmu_device *smmu = dev_get_drvdata(dev);
+	int ret;
 
 	dev_dbg(dev, "Resuming\n");
-	arm_smmu_device_reset(smmu);
+	ret = arm_smmu_device_reset(smmu);
+	if (ret)
+		return ret;
+
+	ret = arm_smmu_enable_irqs(smmu);
+	if (ret) {
+		dev_err(dev, "failed to enable irqs on resume\n");
+		return ret;
+	}
 
 	return 0;
 }
-- 
2.43.0

